symbolic-base/src/ZkFold/Symbolic/Data/UInt.hs

```diff
@@ -74,108 +70,100 @@ import           ZkFold.Symbolic.Data.Ord
 import           ZkFold.Symbolic.Interpreter       (Interpreter (..))
 import           ZkFold.Symbolic.MonadCircuit
 
-
 -- TODO (Issue #18): hide this constructor
-newtype UInt (n :: Natural) (r :: RegisterSize) (context :: (Type -> Type) -> Type) = UInt (context (Vector (NumberOfRegisters (BaseField context) n r)))
+newtype UInt (r :: Natural) (n :: Natural) (c :: (Type -> Type) -> Type) = UInt (c (Vector n))
 
-deriving instance Generic (UInt n r context)
-deriving instance HNFData context => NFData (UInt n r context)
-deriving instance HEq context => Haskell.Eq (UInt n r context)
-deriving instance HShow context => Haskell.Show (UInt n r context)
-deriving newtype instance (KnownRegisters c n r, Symbolic c) => SymbolicData (UInt n r c)
-deriving newtype instance (KnownRegisters c n r, Symbolic c) => Conditional (Bool c) (UInt n r c)
-deriving newtype instance (KnownRegisters c n r, Symbolic c) => Eq (UInt n r c)
+deriving instance Generic (UInt r n c)
+deriving instance HNFData c => NFData (UInt r n c)
+deriving instance HEq c => Haskell.Eq (UInt r n c)
+deriving instance HShow c => Haskell.Show (UInt r n c)
+deriving newtype instance (KnownNat n, Symbolic c) => SymbolicData (UInt r n c)
+deriving newtype instance (KnownNat n, Symbolic c) => Conditional (Bool c) (UInt r n c)
+deriving newtype instance (KnownNat n, Symbolic c) => Eq (UInt r n c)
 
 toNative ::
-  forall n r c a.
-  (Symbolic c, KnownNat n, KnownRegisterSize r, BaseField c ~ a) =>
-  (KnownNat (GetRegisterSize a n r)) =>
-  UInt n r c -> FieldElement c
+  forall r n c .
+  (KnownNat r, Symbolic c) =>
+  UInt r n c -> FieldElement c
 toNative (UInt rs) = FieldElement $ symbolicF rs
-  (Par1 . fromConstant . (`vectorToNatural` registerSize @(BaseField c) @n @r))
-  (fmap Par1 . hornerW @(GetRegisterSize a n r) . toList)
+  (Par1 . fromConstant . (`vectorToNatural` value @r))
+  (fmap Par1 . hornerW @r . toList)
 
-instance (Symbolic c, KnownNat n, KnownRegisterSize r) => FromConstant Natural (UInt n r c) where
-    fromConstant c = UInt . embed @c $ naturalToVector @c @n @r c
+instance (KnownNat r, KnownNat n, Symbolic c, IsValidRegister r n c) => FromConstant Natural (UInt r n c) where
+    fromConstant c = UInt . embed @c $ naturalToVector @r @n @c c
 
-instance (Symbolic c, KnownNat n, KnownRegisterSize r) => FromConstant Integer (UInt n r c) where
-    fromConstant = fromConstant . naturalFromInteger . (`Haskell.mod` (2 ^ getNatural @n))
+-- TODO: This one throws an underflow error if the value is negative. Should we do something about it?
```

Comment:
I intended `FromConstant` to be lawful homomorphisms. So `FromConstant Integer (UInt r n c)` from this PoV should be how it was defined before -- take residue modulo `2 ^ (r * n)`, which is nonnegative, and use `FromConstant Natural` instance.
For non-overflowing conversions, there's `strictConv`